File: mobile/src/main/java/treehou/se/habit/util/MenuTintUtils.java
package treehou.se.habit.util;


import android.content.Context;
import android.content.res.Resources;
import android.graphics.drawable.Drawable;
import android.support.annotation.ColorInt;
import android.support.v4.graphics.drawable.DrawableCompat;
import android.util.TypedValue;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.ImageView;

public class MenuTintUtils {
    private MenuTintUtils() {
    }

    public static void tintAllIcons(Context context, Menu menu) {
        tintAllIcons(menu, getMenuColor(context));
    }

    public static void tintAllIcons(Menu menu, final int color) {
        for (int i = 0; i < menu.size(); ++i) {
            final MenuItem item = menu.getItem(i);
            tintMenuItemIcon(color, item);
            tintShareIconIfPresent(color, item);
        }
    }

    public static @ColorInt int getMenuColor(Context context){
        TypedValue typedValue = new TypedValue();
        Resources.Theme theme = context.getTheme();
        theme.resolveAttribute(android.R.attr.textColorPrimary, typedValue, true);
        return typedValue.data;
    }

    private static void tintMenuItemIcon(int color, MenuItem item) {
        final Drawable drawable = item.getIcon();
        if (drawable != null) {
            final Drawable wrapped = DrawableCompat.wrap(drawable);
            drawable.mutate();
            DrawableCompat.setTint(wrapped, color);
            item.setIcon(drawable);
        }
    }

    private static void tintShareIconIfPresent(int color, MenuItem item) {
        if (item.getActionView() != null) {
            final View actionView = item.getActionView();
            final View expandActivitiesButton = actionView.findViewById(android.support.v7.appcompat.R.id.expand_activities_button);
            if (expandActivitiesButton != null) {
                final ImageView image = (ImageView) expandActivitiesButton.findViewById(android.support.v7.appcompat.R.id.image);
                if (image != null) {
                    final Drawable drawable = image.getDrawable();
                    final Drawable wrapped = DrawableCompat.wrap(drawable);
                    drawable.mutate();
                    DrawableCompat.setTint(wrapped, color);
                    image.setImageDrawable(drawable);
                }
            }
        }
    }
}

File: mobile/src/main/java/treehou/se/habit/util/MenuTintUtils.kt
package treehou.se.habit.util


import android.content.Context
import android.content.res.Resources
import android.graphics.drawable.Drawable
import android.support.annotation.ColorInt
import android.support.v4.graphics.drawable.DrawableCompat
import android.util.TypedValue
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.ImageView

object MenuTintUtils {

    fun tintAllIcons(context: Context, menu: Menu) {
        tintAllIcons(menu, getMenuColor(context))
    }

    fun tintAllIcons(menu: Menu, color: Int) {
        for (i in 0 until menu.size()) {
            val item = menu.getItem(i)
            tintMenuItemIcon(color, item)
            tintShareIconIfPresent(color, item)
        }
    }

    @ColorInt
    fun getMenuColor(context: Context): Int {
        val typedValue = TypedValue()
        val theme = context.theme
        theme.resolveAttribute(android.R.attr.textColorPrimary, typedValue, true)
        return typedValue.data
    }

    private fun tintMenuItemIcon(color: Int, item: MenuItem) {
        val drawable = item.icon
        if (drawable != null) {
            val wrapped = DrawableCompat.wrap(drawable)
            drawable.mutate()
            DrawableCompat.setTint(wrapped, color)
            item.icon = drawable
        }
    }

    private fun tintShareIconIfPresent(color: Int, item: MenuItem) {
        if (item.actionView != null) {
            val actionView = item.actionView
            val expandActivitiesButton = actionView.findViewById<View>(android.support.v7.appcompat.R.id.expand_activities_button)
            if (expandActivitiesButton != null) {
                val image = expandActivitiesButton.findViewById<View>(android.support.v7.appcompat.R.id.image) as ImageView
                val drawable = image.drawable
                val wrapped = DrawableCompat.wrap(drawable)
                drawable.mutate()
                DrawableCompat.setTint(wrapped, color)
                image.setImageDrawable(drawable)
            }
        }
    }
}
